Codes/Visualization/ConstructOrderBookVisualization_20200214.R

```diff
@@ -45,41 +61,95 @@ mydata$clocktime<-format(as.POSIXct((mydata$time),origin="1970-01-01",tz="UTC"),
-rm(mydata)
+#rm(mydata)
 
-##TEST TIME: 10:30am
+##RUN THROUGH LOOP
+min<-1:60 #snapshot every minute for 1 hour
 
+##SET UP EARLIEST INTERVAL TIME (for first loop, this is the beginning of the trading day)
+oldtesttime<-min(mydata$time)
+##FIRST INTERVAL ("TEST TIME"): 10:30am
 testtime<-10.5*60*60
-
-#find submissions at or before this timestamp
-subt<-sub[which(sub$time<=testtime),]
+##INPUT MATRICES
+cols<-c("size","mpid","price","direction","interval")
+subplot<-data.frame(matrix(NA,0,length(cols))); colnames(subplot)<-cols
+mids<-data.frame(cbind(1:length(min),  rep(NA,length(min)))); colnames(mids)<-c("interval","mid")
+subold<-data.frame(matrix(NA,0,length(Cols))); colnames(subold)<-Cols
+xlabels<-rep(NA,length(min))
+
+for (y in 1:length(min)){
+
+#find state of the order book at closest timestap before testtime
+orbk<-mydata[mydata$time<=testtime,]; orbk<-orbk[nrow(orbk),c(1,8:11)]; orbk$mid<-(orbk$askprice1+orbk$bidprice1)/2
+mids$mid[y]<-orbk$mid
+#find submissions before this interval but after previous interval
+subt<-sub[which(sub$time<testtime&sub$time>=oldtesttime),]
+subt<-rbind(subold,subt) #add orders from previous interval
 #find orders that have been executed or cancelled prior to timestamp
-execanct<-execanc[which(execanc$orderID%in%subt$orderID&execanc$time<=testtime),]
-#sum across executions and cancellations (may be several partial or combo)
-totalect<-aggregate(execanct$size,by=list(Category=execanct$orderID),FUN=sum); colnames(totalect)<-c("orderID","size")
-#discard orders that have been completely executed/cancelled by timestamp
-i<-match(subt$orderID,totalect$orderID)
-subt$execanc<-totalect$size[i]; rm(i) #this is the amount of the submitted order that has been executed or cancelled as of timestamp
-subt$execanc<-replace(subt$execanc,is.na(subt$execanc)==1,0) #replace missing with 0
-subt$remainder<-subt$size-subt$execanc #this is the amount of the order remaining as of the timestamp
-  #q<-which(subt$remainder<0) #sanity check
+execanct<-execanc[which(execanc$orderID%in%subt$orderID&execanc$time<testtime&execanc$time>=oldtesttime),]
+      #sum across executions and cancellations (may be several partial or combo)
+      if (nrow(execanct)!=0){
+      totalect<-aggregate(execanct$size,by=list(Category=execanct$orderID),FUN=sum); colnames(totalect)<-c("orderID","size")
+      #discard orders that have been completely executed/cancelled by timestamp
+      i<-match(subt$orderID,totalect$orderID)
+      subt$execanc<-totalect$size[i]; rm(i) #this is the amount of the submitted order that has been executed or cancelled as of timestamp
+      subt$execanc<-replace(subt$execanc,is.na(subt$execanc)==1,0) #replace missing with 0
+      subt$remainder<-subt$size-subt$execanc #this is the amount of the order remaining as of the timestamp
+      if (length(q<-which(subt$remainder<0))!=0){stop("Negative remainders")} #sanity check
+      rm(execanct,totalect)
+      }
+      else {subt$remainder<-subt$size}
```

Comment:
I fixed a mistake here. In R, the "else" function needs to be on the same line as the closing "}" from the if statement.